rpc/src/v1/impls/eth_filter.rs

```diff
@@ -246,11 +270,75 @@ impl<T: Filterable + Send + Sync + 'static> EthFilter for T {
 
 					// retrieve logs in range from_block..min(BlockId::Latest..to_block)
 					let limit = filter.limit;
-					Either::B(self.logs(filter)
+					Either::B(Either::A(self.logs(filter)
 						.map(move |logs| { reorg.extend(logs); reorg }) // append reorg logs in the front
 						.map(move |mut logs| { logs.extend(pending); logs }) // append fetched pending logs
 						.map(move |logs| limit_logs(logs, limit)) // limit the logs
-						.map(FilterChanges::Logs))
+						.map(FilterChanges::Logs)))
+				},
+				PollFilter::ReturnData(ref mut block_number) => {
+					// +1, cause we want to return hashes including current block hash.
+					let current_number = self.best_block_number() + 1;
+					let executed: Vec<(BlockId, Box<Vec<Bytes>>)> = (*block_number..current_number)
+						.filter_map(|block| {
+							let block_id = BlockId::Number(block);
+							let replay_result = self.replay_block_transactions(block_id);
+							match replay_result {
+								Ok(Ok(executeds)) => {
+									let output_bytes = Box::new(executeds
+																.map(|executed| executed.output)
+																.map(Bytes::from)
+																.collect()
+															   );
+									Some((block_id, output_bytes))
+								},
+								Ok(Err(e)) => {
+									warn!("Error replaying transactions for block {:?}: {:?}", block_id, e);
+									None
+								},
+								Err(e) => {
+									warn!("Error replaying transactions for block {:?}: {:?}", block_id, e);
+									None
+								},
+							}
+						})
+					.collect();
+					let return_data: Vec<_> = executed
+						.into_iter()
+						.map(|(block_id, output_bytes)| {
+							self.block_body(block_id)
+								.map(|body| {
+									match body {
+										None => vec![],
```

Comment:
What if block body does not exist? Is it possible? Later on you do `flat_map` on the result. If it's possible for a block body to not exist, the function will produce incorrect results. If it always exist, please write `.expect` here with a proof